core/src/test/java/org/elasticsearch/common/netty/NettyUtilsTests.java

```
@@ -0,0 +1,107 @@
+        assertArrayEquals(slice.toBytes(), bytesReference.toBytes());
+    }
+
+    public void testToChannelBufferWithSliceAfter() {
+        BytesReference ref = getRandomizedBytesReference(randomIntBetween(1, 3 * PAGE_SIZE));
+        int sliceOffset = randomIntBetween(0, ref.length());
+        int sliceLength = randomIntBetween(ref.length() - sliceOffset, ref.length() - sliceOffset);
+        ChannelBuffer channelBuffer = NettyUtils.toChannelBuffer(ref);
+        BytesReference bytesReference = NettyUtils.toBytesReference(channelBuffer);
+        assertArrayEquals(ref.slice(sliceOffset, sliceLength).toBytes(), bytesReference.slice(sliceOffset, sliceLength).toBytes());
+    }
+
+    public void testToChannelBuffer() {
+        BytesReference ref = getRandomizedBytesReference(randomIntBetween(1, 3 * PAGE_SIZE));
+        ChannelBuffer channelBuffer = NettyUtils.toChannelBuffer(ref);
+        BytesReference bytesReference = NettyUtils.toBytesReference(channelBuffer);
+        if (ref instanceof ChannelBufferBytesReference) {
+            assertEquals(channelBuffer, ((ChannelBufferBytesReference) ref).toChannelBuffer());
+        } else if (ref.hasArray() == false) { // we gather the buffers into a channel buffer
+            assertTrue(channelBuffer instanceof CompositeChannelBuffer);
+        }
+        assertArrayEquals(ref.toBytes(), bytesReference.toBytes());
+    }
+
+    private BytesReference getRandomizedBytesReference(int length) {
+        // TODO we should factor out a BaseBytesReferenceTestCase
+        // we know bytes stream output always creates a paged bytes reference, we use it to create randomized content
+        ReleasableBytesStreamOutput out = new ReleasableBytesStreamOutput(length, bigarrays);
+        try {
+            for (int i = 0; i < length; i++) {
+                out.writeByte((byte) random().nextInt(1 << 8));
+            }
+        } catch (IOException e) {
+            fail("should not happen " + e.getMessage());
```

should we just let the exception propagate?